ocaml/typing/typedtree.ml

```diff
@@ -51,15 +51,42 @@ type _ pattern_category =
 | Computation : computation pattern_category
 
 module Unique_barrier = struct
-  type t = Mode.Uniqueness.lr
-
-  let fresh () = Uniqueness.newvar ()
-
+  (* For added safety, we record the states in the life of a barrier:
+     - Barriers start out as Not_computed
+     - They are enabled by the uniqueness analysis
+     - They are resolved in translcore
+    This allows us to error if the barrier is not enabled or resolved. *)
+  type barrier =
+    | Enabled of Mode.Uniqueness.lr
+    | Resolved of Mode.Uniqueness.Const.t
+    | Not_computed
+
+  type t = barrier ref
+
+  let not_computed () = ref Not_computed
+
+  let enable barrier = match !barrier with
+    | Not_computed ->
+      barrier := Enabled (Uniqueness.newvar ())
+    | _ -> assert false
+
+  (* Due to or-patterns a barrier may have several upper bounds. *)
   let add_upper_bound uniq barrier =
-    Uniqueness.submode_exn barrier uniq
+    match !barrier with
+    | Enabled barrier -> Uniqueness.submode_exn barrier uniq
+    | _ -> assert false
 
   let resolve barrier =
-    Uniqueness.zap_to_ceil barrier
+    match !barrier with
+    | Enabled uniq ->
+      let zapped = Uniqueness.zap_to_ceil uniq in
+      barrier := Resolved zapped;
+      zapped
+    | Resolved barrier -> barrier
```

Comment:
Does this case ever happen? If not, we don't need `Resolved`.

Reply:
It can now happen after your suggestion below :) It also enables us to write the check that you proposed, where we traverse all unique barriers after generating lambda to check that all were resolved.

Reply:
Ah, yes, that makes sense. OK.